### PR TITLE
layers: Adds shader validation for VK_NV_shader_image_footprint

### PR DESCRIPTION
**Pull request by jmacnak-nv**
Adds one more shader validation check for VK_NV_shader_image_footprint and adds a couple tests for the recent shader validation checks that were added.

### CODE REVIEW

**Comment by jmacnak-nv** [comment]
(just added another commit to remove extra new line, sorry!)

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -2539,6 +2539,12 @@ void CoreChecks::PostCallRecordCreateDevice(VkPhysicalDevice gpu, const VkDevice
         core_checks->enabled_features.fragment_shader_barycentric_features = *fragment_shader_barycentric_features;
     }
 
+    const auto *shader_image_footprint_features =
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp]: @mark-lunarg @mikes-lunarg -- we're up to 17 instances of this pattern, each of which is a linear search of the list 17*(n/2).  Even for a full list -- 17*17/2 (< 150) linked list if check and traverse,  I'm not concerned about performance, but we should look at replacing this with a template function.

File: tests/vklayertests_pipeline_shader.cpp [hunk 2]
```diff
@@ -6886,3 +6886,185 @@ TEST_F(VkLayerTest, NVRayTracingPipelineStageCreationFeedbackCount) {
...
+    // Disable the image footprint feature.
+    VkPhysicalDeviceShaderImageFootprintFeaturesNV image_footprint_features = {};
+    image_footprint_features.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SHADER_IMAGE_FOOTPRINT_FEATURES_NV;
+    image_footprint_features.imageFootprint = VK_FALSE;
+
+    void *pnext_chain = reinterpret_cast<void *>(&image_footprint_features);
+
+    VkDeviceObj test_device(0, gpu(), device_extension_names, &features, pnext_chain);
+
+    char const *vsSource =
```
Comment (jzulauf-lunarg) [inline tests/vklayertests_pipeline_shader.cpp]: @locke-lunarg -- is there anything here that should reflect the new best practices, predefined shaders, etc of your latest refactor

File: tests/vklayertests_pipeline_shader.cpp [hunk 3]
```diff
@@ -6886,3 +6886,185 @@ TEST_F(VkLayerTest, NVRayTracingPipelineStageCreationFeedbackCount) {
...
+    TEST_DESCRIPTION("Create a pipeline requiring the shader image footprint feature which has not enabled on the device.");
+
+    ASSERT_NO_FATAL_FAILURE(Init());
+    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
+
+    std::vector<const char *> device_extension_names;
+    auto features = m_device->phy().features();
+
+    // Disable the image footprint feature.
+    VkPhysicalDeviceShaderImageFootprintFeaturesNV image_footprint_features = {};
```
Comment (jzulauf-lunarg) [inline tests/vklayertests_pipeline_shader.cpp]: See 'lvl_init_struct'  usage.  Slightly terser, with typesafe sType auto initialization.

File: tests/vklayertests_pipeline_shader.cpp [hunk 4]
```diff
@@ -6886,3 +6886,185 @@ TEST_F(VkLayerTest, NVRayTracingPipelineStageCreationFeedbackCount) {
...
+    image_footprint_features.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SHADER_IMAGE_FOOTPRINT_FEATURES_NV;
+    image_footprint_features.imageFootprint = VK_FALSE;
+
+    void *pnext_chain = reinterpret_cast<void *>(&image_footprint_features);
```
Comment (jzulauf-lunarg) [inline tests/vklayertests_pipeline_shader.cpp]: The void * cast and local are unneeded, just pass the &image_footprint_features

File: tests/vklayertests_pipeline_shader.cpp [hunk 5]
```diff
@@ -6886,3 +6886,185 @@ TEST_F(VkLayerTest, NVRayTracingPipelineStageCreationFeedbackCount) {
...
+    TEST_DESCRIPTION("Create a pipeline requiring the fragment shader barycentric feature which has not enabled on the device.");
+
+    ASSERT_NO_FATAL_FAILURE(Init());
+    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
+
+    std::vector<const char *> device_extension_names;
+    auto features = m_device->phy().features();
+
+    // Disable the fragment shader barycentric feature.
+    VkPhysicalDeviceFragmentShaderBarycentricFeaturesNV fragment_shader_barycentric_features = {};
```
Comment (jzulauf-lunarg) [inline tests/vklayertests_pipeline_shader.cpp]: See lvl_init and void * cast comments above

File: tests/vklayertests_pipeline_shader.cpp [hunk 6]
```diff
@@ -6886,3 +6886,185 @@ TEST_F(VkLayerTest, NVRayTracingPipelineStageCreationFeedbackCount) {
...
+    fragment_shader_barycentric_features.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FRAGMENT_SHADER_BARYCENTRIC_FEATURES_NV;
+    fragment_shader_barycentric_features.fragmentShaderBarycentric = VK_FALSE;
+
+    void *pnext_chain = reinterpret_cast<void *>(&fragment_shader_barycentric_features);
+
+    VkDeviceObj test_device(0, gpu(), device_extension_names, &features, pnext_chain);
+
+    char const *vsSource =
```
Comment (jzulauf-lunarg) [inline tests/vklayertests_pipeline_shader.cpp]: Duplicates `bindStateVertShaderText`. Delete and reference this static const instead.

File: tests/vklayertests_pipeline_shader.cpp (hunk 2 above)
Comment (jzulauf-lunarg) [reply tests/vklayertests_pipeline_shader.cpp]: Duplicates `bindStateVertShaderText`. Delete and reference this static const instead.

File: tests/vklayertests_pipeline_shader.cpp [hunk 7]
```diff
@@ -6886,3 +6886,185 @@ TEST_F(VkLayerTest, NVRayTracingPipelineStageCreationFeedbackCount) {
...
+
+TEST_F(VkLayerTest, CreatePipelineCheckComputeShaderDerivativesEnabled) {
+    TEST_DESCRIPTION("Create a pipeline requiring the compute shader derivatives feature which has not enabled on the device.");
+
+    ASSERT_NO_FATAL_FAILURE(Init());
+
+    std::vector<const char *> device_extension_names;
+    auto features = m_device->phy().features();
+
+    // Disable the compute shader derivatives features.
```
Comment (jzulauf-lunarg) [inline tests/vklayertests_pipeline_shader.cpp]: lvl_init / void * cast comments ...

File: tests/vklayertests_pipeline_shader.cpp [hunk 8]
```diff
@@ -6886,3 +6886,185 @@ TEST_F(VkLayerTest, NVRayTracingPipelineStageCreationFeedbackCount) {
...
+                                        0,
+                                        {VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO, nullptr, 0,
+                                         VK_SHADER_STAGE_COMPUTE_BIT, cs.handle(), "main", nullptr},
+                                        pl.handle(),
+                                        VK_NULL_HANDLE,
+                                        -1};
+
+    m_errorMonitor->SetDesiredFailureMsg(
+        VK_DEBUG_REPORT_ERROR_BIT_EXT,
+        "Shader requires VkPhysicalDeviceComputeShaderDerivativesFeaturesNV::computeDerivativeGroupQuads but is not enabled on the "
```
Comment (jzulauf-lunarg) [inline tests/vklayertests_pipeline_shader.cpp]: (Not required for this PR)

@mikes-lunarg this whole class of error messages seems ripe for a) a set of "UNASSIGNED-"  which would then be part of the UNASSIGNED -> VUID task.

**Comment by jzulauf-lunarg** [comment]
@jeremy-lunarg  -- shader_validation review
@locke-lunarg  -- unit test best practices review.

File: tests/vklayertests_pipeline_shader.cpp (hunk 2 above)
Comment (jmacnak-nv) [reply tests/vklayertests_pipeline_shader.cpp]: Addressed the bindStateVertShaderText comment.

File: tests/vklayertests_pipeline_shader.cpp (hunk 3 above)
Comment (jmacnak-nv) [reply tests/vklayertests_pipeline_shader.cpp]: Done.

File: tests/vklayertests_pipeline_shader.cpp (hunk 4 above)
Comment (jmacnak-nv) [reply tests/vklayertests_pipeline_shader.cpp]: Done.

File: tests/vklayertests_pipeline_shader.cpp (hunk 5 above)
Comment (jmacnak-nv) [reply tests/vklayertests_pipeline_shader.cpp]: Done.

File: tests/vklayertests_pipeline_shader.cpp (hunk 6 above)
Comment (jmacnak-nv) [reply tests/vklayertests_pipeline_shader.cpp]: Done.

File: tests/vklayertests_pipeline_shader.cpp (hunk 7 above)
Comment (jmacnak-nv) [reply tests/vklayertests_pipeline_shader.cpp]: Done.

**Comment by mark-lunarg** [comment]
OK, I'll pull it and run it through CI.

**Comment by mark-lunarg** [comment]
Rebased, got a clean pass through CI, reformatted commit messages and pushed to Master.  Thanks!

**Pull request closed without merging**